src/ast/ast_rewrite_star_projections.c

```diff
@@ -0,0 +1,415 @@
+			cypher_astnode_t *identifier = cypher_ast_identifier(name,
+																 strlen(name), range);
+			array_append(*aliases, identifier);
+		}
+		Proc_Free(proc);
+		return;
+	}
+
+	for(uint i = 0; i < yield_count; i ++) {
+		const cypher_astnode_t *projection = cypher_ast_call_get_projection(call_clause, i);
+		const cypher_astnode_t *ast_exp = cypher_ast_projection_get_expression(projection);
+
+		const cypher_astnode_t *alias_node = cypher_ast_projection_get_alias(projection);
+		if(alias_node == NULL) alias_node = ast_exp;
+
+		array_append(*aliases, alias_node);
+	}
+}
+
+static const cypher_astnode_t **_collect_aliases_in_scope
+(
+	const cypher_astnode_t *root,
+	uint scope_start,
+	uint scope_end
+) {
+	ASSERT(scope_start != scope_end);
+	const cypher_astnode_t **aliases = array_new(const cypher_astnode_t *, 1);
+
+	for(uint i = scope_start; i < scope_end; i ++) {
+		const cypher_astnode_t *clause = cypher_ast_query_get_clause(root, i);
+		cypher_astnode_type_t type = cypher_astnode_type(clause);
+
+		if(type == CYPHER_AST_WITH) {
+			// the WITH clause contains either
+			// aliases or its own STAR projection
+			_collect_with_projections(clause, &aliases);
+		} else if(type == CYPHER_AST_MATCH) {
+			// the MATCH clause contains one pattern of N paths
+			const cypher_astnode_t *pattern =
+				cypher_ast_match_get_pattern(clause);
+			_collect_aliases_in_pattern(pattern, &aliases);
+		} else if(type == CYPHER_AST_CREATE) {
+			// the CREATE clause contains one pattern of N paths
+			const cypher_astnode_t *pattern =
+				cypher_ast_create_get_pattern(clause);
+			_collect_aliases_in_pattern(pattern, &aliases);
+		} else if(type == CYPHER_AST_MERGE) {
+			// the MERGE clause contains one path
+			const cypher_astnode_t *path =
+				cypher_ast_merge_get_pattern_path(clause);
+			_collect_aliases_in_path(path, &aliases);
+		} else if(type == CYPHER_AST_UNWIND) {
+			// the UNWIND clause introduces one alias
+			const cypher_astnode_t *unwind_alias =
+				cypher_ast_unwind_get_alias(clause);
+			array_append(aliases, unwind_alias);
+		} else if(type == CYPHER_AST_CALL) {
+			_collect_call_projections(clause, &aliases);
+		}
+	}
+
+	// sort and unique the aliases array
+	// so that we won't make redundant projections given queries like:
+	// MATCH (a)-[]->(a) RETURN *
+	_uniqueArray(aliases);
+
+	return aliases;
+}
+
+static void replace_clause
+(
+	cypher_astnode_t *root,    // ast root
+	cypher_astnode_t *clause,  // clause being replaced
+	int scope_start,           // begining of scope
+	int scope_end              // ending of scope
+) {
+	cypher_astnode_type_t t = cypher_astnode_type(clause);
+
+	//--------------------------------------------------------------------------
+	// collect identifiers
+	//--------------------------------------------------------------------------
+
+	// TODO: MATCH (a)-[]->(a) RETURN *
+	// 'a' is duplicated ?
+	const cypher_astnode_t **identifiers = _collect_aliases_in_scope(root,
```

Comment:
Will `MATCH (a)-[]->(a) RETURN *` introduce `a` as a duplicate?

Reply:
No, it will only be introduced once:

https://github.com/RedisGraph/RedisGraph/pull/2147/commits/a86b8e5f766963834c3815895ae5eeb6cb51c0f5#diff-3b94fdb36b9558c999b6fb398c7b39f047e1fafdf5ef00879521e25277244ecbR182-R185